ludwig/models/trainer.py

```diff
@@ -91,40 +94,84 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         self.shutdown()
 
 
+@dataclass
+class TrainerConfig:
+    optimizer: Dict = schema.OptimizerOptions(default={TYPE: "adam"})
+    epochs: int = schema.PositiveInteger(default=100)
+    regularization_lambda: float = schema.FloatRange(default=0.0, min=0)
+    regularization_type: Optional[str] = schema.StringOptions(
+        options=["l1", "l2", "l1_l2"], default="l2", nullable=True
+    )
+    learning_rate: float = schema.NumericOrStringOptionsField(
+        default=0.001, min=0.0, max=1.0, options=["auto"], nullable=False
+    )
+    batch_size: Union[int, str] = schema.IntegerOrStringOptionsField(
+        default=128, options=["auto"], nullable=False, exclusiveMin=0
+    )
+    eval_batch_size: Optional[int] = schema.IntegerOrStringOptionsField(
+        default=None, options=["auto"], nullable=True, exclusiveMin=0
+    )
+    early_stop: int = schema.IntegerRange(default=5, min=-1)  # Can be -1, which disables early_stop
+    reduce_learning_rate_on_plateau: float = schema.FloatRange(default=0.0, min=0.0, max=1.0)
+    reduce_learning_rate_on_plateau_patience: int = schema.NonNegativeInteger(5)
+    reduce_learning_rate_on_plateau_rate: float = schema.FloatRange(default=0.5, min=0.0, max=1.0)
+    increase_batch_size_on_plateau: int = schema.NonNegativeInteger(default=0)
+    increase_batch_size_on_plateau_patience: int = schema.NonNegativeInteger(default=5)
+    increase_batch_size_on_plateau_rate: float = schema.FloatRange(default=2.0, min=0.0)
+    increase_batch_size_on_plateau_max: int = schema.PositiveInteger(default=512)
+    decay: bool = False
+    decay_steps: int = schema.PositiveInteger(default=10000)
+    decay_rate: float = schema.FloatRange(default=0.96, min=0.0, max=1.0)
+    staircase: bool = False
+    # "gradient_clipping": None, # TODO: What is this?
```

Comment:
[Gradient clipping](https://deepai.org/machine-learning-glossary-and-terms/gradient-clipping#:~:text=Gradient%20clipping%20is%20a%20technique,input%20into%20a%20specific%20output.) is the practice of forcing the gradient values (element-wise) to a specific minimum or maximum value if the gradient exceeded an expected range. 

It appears that this might have been replaced by the Clipper data class, used by `create_optimizer_with_clipper`, which is configured using the `**optimizer` config. Should we add the attributes of the `Clipper` dataclass to the schema for `OptimizerOptions`?

Reply:
For stronger 'typing' I was actually going to make small dataclasses for each optimizer type and remove `OptimizerOptions`. 

Is `Clipper` functionality always going to be tied to optimizer configuration? (Seems to be the case currently.) If so, then yeah I think we might as well move those attributes over, but if it's worth keeping `Clipper` as a separate dataclass then I could just add a `clipperConfig` attribute to each optimizer dataclass that is `None` by default.

Reply:
I could go either way on whether to keep Clipper as a separate dataclass vs. copying each attribute over to each optimizer.

Reply:
clipper is being kept as a separate dataclass and is now explicitly on the trainer dataclass, `create_optimizer_with_clipper` has been updated